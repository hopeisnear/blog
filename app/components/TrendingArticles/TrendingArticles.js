import React from 'react';
import './TrendingArticles.scss';

const TrendingArticles = () => (
  <div className="trending-articles">
    <div className="header">
      Trending articles
    </div>
    <div className="content">
      <nav className="blog-block">
        <ul className="blog_categories-list">
          <li className="blog_categories-list_item">
            <a href="https://www.toptal.com/developers/blog/back-end">
                Back-end
            </a>
          </li>
          <li className="blog_categories-list_item">
            <a href="https://www.toptal.com/developers/blog/web-front-end">
                Web front-end
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </div>
);

export default TrendingArticles;
