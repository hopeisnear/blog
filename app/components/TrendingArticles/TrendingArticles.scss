@import 'app/styles/variables.scss';

.trending-articles {
  .header {
    font-family: $secondary-font;
    letter-spacing: $secondary-font-letter-spacing;
    font-size: $secondary-font-size;
    font-weight: $secondary-font-weight;
    line-height: $secondary-font-line-height;
    color: $secondary-color;
    text-transform: uppercase;
    margin-bottom: 15px;
  }
  .content {
    background: $secondary-background-color;
    box-shadow: $main-box-shadow;
    border: 1px solid $main-border-color;

    .trending-articles-link {
      display: flex;
      flex-direction: row;
      flex-grow: 1;
      flex-basis: 100%;
      border-bottom: 1px solid $secondary-border-color;
      .image {
        flex-shrink: 0;
        width: 75px;
        min-height: 110px;
        display: block;
        overflow: hidden;
        position: relative;
      }
      .trending-articles-desciption {
        font-family: $secondary-font;
        font-size: $font-size-sm;
        line-height: $font-size-sm-line-height;
        padding: 10px 20px 10px 15px;
      }
      &:hover {
        .image {
          img.content-image {
            opacity: 1;
            width: 550%;
          }
        }
      }
    }
  }
}
