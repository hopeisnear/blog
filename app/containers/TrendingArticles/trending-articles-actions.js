import { fetchTrendingArticles } from 'services/article-service';
import { FETCH_TRENDING_ARTICLES_RESPONDED } from 'common/article-actions';

export function fetchTrendingArticlesAction() {
  return dispatch =>
    fetchTrendingArticles().then(trendingArticles => {
      dispatch({ type: FETCH_TRENDING_ARTICLES_RESPONDED, trendingArticles });
    });
}
