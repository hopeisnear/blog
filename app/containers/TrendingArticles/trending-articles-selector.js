import { createSelector } from 'reselect';

const selectTrendingArticles = () => createSelector(state => state.get('blog'), blog => blog.get('trendingArticles'));

export { selectTrendingArticles };
