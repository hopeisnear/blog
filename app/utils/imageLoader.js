// eslint-disable-next-line global-require
// eslint-disable-next-line import/no-dynamic-require,global-require
export const loadImage = (imageName) => require(`images/${imageName}`);
