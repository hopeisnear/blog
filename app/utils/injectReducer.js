import React from 'react';
import { shape } from 'prop-types';
import hoistNonReactStatics from 'hoist-non-react-statics';

import getInjectors from './reducerInjectors';

export default ({ key, reducer }) => (WrappedComponent) => {
  class ReducerInjector extends React.Component {
    static contextTypes = {
      store: shape({}).isRequired
    };

    injectors = getInjectors(this.context.store);

    componentWillMount() {
      const { injectReducer } = this.injectors;

      injectReducer(key, reducer);
    }

    render() {
      return <WrappedComponent {...this.props} />;
    }
  }

  return hoistNonReactStatics(ReducerInjector, WrappedComponent);
};
