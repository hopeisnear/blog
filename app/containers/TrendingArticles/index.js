import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';
import { fetchTrendingArticlesAction } from './trending-articles-actions';
import { selectTrendingArticles } from './trending-articles-selector';
import TrendingArticles from './TrendingArticles';

const mapDispatchToProps = dispatch => ({
  fetchTrendingArticles: () => dispatch(fetchTrendingArticlesAction())
});

const mapStateToProps = createStructuredSelector({
  trendingArticles: selectTrendingArticles()
});

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(withConnect)(TrendingArticles);
export { mapDispatchToProps };
