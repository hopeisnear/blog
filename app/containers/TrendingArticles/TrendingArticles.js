import React, { PureComponent } from 'react';
import { arrayOf, shape, func } from 'prop-types';
import { Link } from 'react-router-dom';
import { imageUrlFor } from 'utils/imageLoader';
import './TrendingArticles.scss';

export default class TrendingArticles extends PureComponent {
  componentDidMount() {
    this.props.fetchTrendingArticles();
  }

  render() {
    return (
      <div className="TrendingArticles">
        <div className="trending-articles__header">Trending articles</div>
        <div className="trending-articles__body">
          <nav>
            {this.props.trendingArticles.map(trendingArticle => (
              <Link key={trendingArticle.slug.current} className="trending-articles__link" to={`/blog/${trendingArticle.slug.current}`}>
                <span className="trending-articles__image">
                  <img
                    src={imageUrlFor(trendingArticle.image)
                      .ignoreImageParams()
                      .width(360)
                      .height(252)}
                    alt="article heading"
                  />
                </span>
                <span className="trending-articles__title">{trendingArticle.title}</span>
              </Link>
            ))}
          </nav>
        </div>
      </div>
    );
  }
}

TrendingArticles.propTypes = {
  trendingArticles: arrayOf(shape({})).isRequired,
  fetchTrendingArticles: func.isRequired
};
