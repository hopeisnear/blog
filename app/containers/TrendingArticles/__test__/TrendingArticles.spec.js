import React from 'react';
import { shallow } from 'enzyme';
import { spy, stub, assert } from 'sinon';
import * as imageLoader from 'utils/imageLoader';
import TrendingArticles from '../TrendingArticles';

describe('<TrendingArticles />', () => {
  test('should render <TrendingArticles />', () => {
    stub(imageLoader, 'imageUrlFor').callsFake(image => ({ ignoreImageParams: () => ({ width: () => ({ height: () => image }) }) }));

    const trendingArticlesComponent = shallow(<TrendingArticles {...createProps()} />);

    expect(trendingArticlesComponent).toMatchSnapshot();
  });

  test('should load trending articles on component init', () => {
    stub(imageLoader, 'imageUrlFor').callsFake(image => ({ ignoreImageParams: () => ({ width: () => ({ height: () => image }) }) }));

    const trendingArticlesComponent = shallow(<TrendingArticles {...createProps()} />);

    assert.calledOnce(trendingArticlesComponent.instance().props.fetchTrendingArticles);
  });

  afterEach(() => {
    imageLoader.imageUrlFor.restore();
  });

  function createProps() {
    return {
      fetchTrendingArticles: spy(),
      trendingArticles: [
        {
          slug: { current: 'article-1' },
          title: 'article-1-title',
          image: 'image1'
        },
        {
          slug: { current: 'article-2' },
          title: 'article-2-title',
          image: 'image2'
        }
      ]
    };
  }
});
