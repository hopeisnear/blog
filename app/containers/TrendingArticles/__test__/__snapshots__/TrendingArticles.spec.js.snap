// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrendingArticles /> should render <TrendingArticles /> 1`] = `
<div
  className="TrendingArticles"
>
  <div
    className="trending-articles__header"
  >
    Trending articles
  </div>
  <div
    className="trending-articles__body"
  >
    <nav>
      <Link
        className="trending-articles__link"
        key="article-1"
        replace={false}
        to="/article-1"
      >
        <span
          className="trending-articles__image"
        >
          <img
            alt="article heading"
            src="image1"
          />
        </span>
        <span
          className="trending-articles__title"
        >
          article-1-title
        </span>
      </Link>
      <Link
        className="trending-articles__link"
        key="article-2"
        replace={false}
        to="/article-2"
      >
        <span
          className="trending-articles__image"
        >
          <img
            alt="article heading"
            src="image2"
          />
        </span>
        <span
          className="trending-articles__title"
        >
          article-2-title
        </span>
      </Link>
    </nav>
  </div>
</div>
`;
